apps/agent/app/src/script.js

```diff
@@ -47,290 +74,463 @@ const retryEvery = async (
+      amount: await loadTokenBalance(tokenAddress),
+    }
+    return newBalances
+  }
+}
+
+async function newBalanceEntry(tokenContract, tokenAddress, settings) {
+  const [balance, decimals, name, symbol] = await Promise.all([
+    loadTokenBalance(tokenAddress),
+    loadTokenDecimals(tokenContract, tokenAddress, settings),
+    loadTokenName(tokenContract, tokenAddress, settings),
+    loadTokenSymbol(tokenContract, tokenAddress, settings),
+  ])
+
+  return {
+    decimals,
+    name,
+    symbol,
+    address: tokenAddress,
+    amount: balance,
+    verified:
+      isTokenVerified(tokenAddress, settings.network.type) ||
+      addressesEqual(tokenAddress, settings.ethToken.address),
+  }
+}
+
+function loadTokenBalance(tokenAddress) {
+  return app.call('balance', tokenAddress).toPromise()
+}
+
+async function loadTokenDecimals(tokenContract, tokenAddress, { network }) {
+  if (tokenDecimals.has(tokenContract)) {
+    return tokenDecimals.get(tokenContract)
+  }
+
+  const fallback =
+    tokenDataFallback(tokenAddress, 'decimals', network.type) || '0'
+
+  let decimals
+  try {
+    decimals = (await tokenContract.decimals().toPromise()) || fallback
+    tokenDecimals.set(tokenContract, decimals)
+  } catch (err) {
+    // decimals is optional
+    decimals = fallback
+  }
+  return decimals
+}
+
+async function loadTokenName(tokenContract, tokenAddress, { network }) {
+  if (tokenNames.has(tokenContract)) {
+    return tokenNames.get(tokenContract)
+  }
+  const fallback = tokenDataFallback(tokenAddress, 'name', network.type) || ''
+
+  let name
+  try {
+    name = (await getTokenName(app, tokenAddress)) || fallback
+    tokenNames.set(tokenContract, name)
+  } catch (err) {
+    // name is optional
+    name = fallback
+  }
+  return name
+}
+
+async function loadTokenSymbol(tokenContract, tokenAddress, { network }) {
+  if (tokenSymbols.has(tokenContract)) {
+    return tokenSymbols.get(tokenContract)
+  }
+  const fallback = tokenDataFallback(tokenAddress, 'symbol', network.type) || ''
+
+  let symbol
+  try {
+    symbol = (await getTokenSymbol(app, tokenAddress)) || fallback
+    tokenSymbols.set(tokenContract, symbol)
+  } catch (err) {
+    // symbol is optional
+    symbol = fallback
+  }
+  return symbol
+}
+
+/***********************
+ *                     *
+ * Transaction Helpers *
+ *                     *
+ ***********************/
+
+async function updateTransactions(transactions, event, detailOptions) {
+  const newTransactions = Array.from(transactions || [])
+
+  const transactionDetails = await createTransactionDetails(
+    event,
+    detailOptions
+  )
+
+  const transactionsIndex = newTransactions.findIndex(
+    ({ id }) => id === transactionDetails.id
+  )
+  if (transactionsIndex === -1) {
+    return newTransactions.concat(transactionDetails)
+  } else {
+    newTransactions[transactionsIndex] = transactionDetails
+    return newTransactions
+  }
+}
+
+async function createTransactionDetails(
```

Comment:
Generally we've called this function "marshall" (as it's converting types from another context)